library/pkcs7.c

```diff
@@ -476,16 +441,46 @@ static int pkcs7_get_signed_data( unsigned char *buf, size_t buflen,
         return( MBEDTLS_ERR_PKCS7_INVALID_ALG );
     }
 
-    /* Do not expect any content */
-    ret = pkcs7_get_content_info_type( &p, end, &signed_data->content.oid );
-    if( ret != 0 )
-        return( ret );
+    /**
+     * ContentInfo ::= SEQUENCE {
+     **/
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                                                   | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
+        return MBEDTLS_ERROR_ADD( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret );
+
+    end_content = p + len;
+
+    /*
+     *      contentType ContentType,
+     *
+     *      ContentType ::= OBJECT IDENTIFIER
+     */
+    if( ( ret = mbedtls_asn1_get_tag( &p, end_content, &len, MBEDTLS_ASN1_OID ) ) != 0 )
+        return MBEDTLS_ERROR_ADD( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret );
+
+    signed_data->content.oid.tag = MBEDTLS_ASN1_OID;
+    signed_data->content.oid.len = len;
+    signed_data->content.oid.p = p;
+    p += len;
 
     if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &signed_data->content.oid ) )
     {
         return( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO );
     }
 
+    /*
+     *      content      [0] EXPLICIT ANY DEFINED BY contentType OPTIONAL }
+     */
+    if( p == end_content )
+        memset( &signed_data->content.data, 0, sizeof( signed_data->content.data ) );
+    else if( ( content_info_ret = pkcs7_get_next_content_len( &p, end_content, &len ) ) == 0 )
+    {
+        signed_data->content.data.tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_CONTEXT_SPECIFIC;
+        signed_data->content.data.len = len;
+        signed_data->content.data.p = p;
+    }
+    p = end_content;
```

Comment:
Assert that `p+len == end_content`, and that `pkcs_get_next_content_len` returned 0.

If either is not true, return an error here.